enhancements/network/ovn-kubernetes-ipsec.md

```diff
@@ -0,0 +1,298 @@
+	// +kubebuilder:validation:Minimum=0
+	// +optional
+	MTU *uint32 `json:"mtu,omitempty"`
+	// geneve port is the UDP port to be used by geneve encapulation.
+	// Default is 6081
+	// +kubebuilder:validation:Minimum=1
+	// +optional
+	GenevePort *uint32 `json:"genevePort,omitempty"`
+	// HybridOverlayConfig configures an additional overlay network for peers that are
+	// not using OVN.
+	// +optional
+  HybridOverlayConfig *HybridOverlayConfig `json:"hybridOverlayConfig,omitempty"`
+  // EnableIPsec enables IPsec encryption on all pod-pod traffic within the cluser. If unset, the plugin will default
+  // EnableIPsec to false.
+  // +optional
+  EnableIPsec *bool `json:"enableIPsec,omitempty"`
+}
+```
+
+This will result in an updated OVNKubernetes Operator configuration object:
+(.spec.defaultNetwork.ovnKubernetesConfig) which will enable cluster-wide
+configuration of IPsec at cluster installation time:
+
+spec:
+  defaultNetwork:
+    type: OVNKubernetes
+    ovnKubernetesConfig:
+      mtu: 1400
+      genevePort: 6081
+      **enableIPsec: true/false**
+
+### User Stories
+
+#### Story 1
+
+As a cluster operator, I want to be able to enable IPsec at cluster installation
+time in the OVNKubernetes plugin in order to encrypt, authenticate and ensure
+the integrity of all pod-pod traffic.
+
+### Implementation Details/Notes/Constraints [optional]
+
+It is expected that the following changes will need to be made in order to
+enable this enhancement:
+
+- OVNKubernetes pod Dockerfiles will need to be updated in order to install
+  necessary components required by OVN/OVS to enable IPsec (e.g.
+  openvswitch-ipsec, libreswan, etc).
+- Modify Cluster Network Operator manifests to configure start up of packages
+  required by OVN/OVS in order to enable IPsec (e.g. ovs-monitor-ipsec, pluto,
+  etc) and determine how to containerize these components.
+- Ensure that all versions of all components are at the correct level required
+  by OVN/OVS. This may required back-porting of patches into RHEL. It has
+  already been identified that OVS and the Kernel will each require a patch to
+  enable this enhancement.
+- Ensure correct generation of the Certificate Authority (CA) certificate and
+  all node private keys and CA-signed certificates. It is envisioned that the
+  key generation mechanism (and keys) of OVN SSL will be reused.
+- Modify Cluster Network Operator manifests to ensure that all OVS/OVN
+  components are configured to use host private key, host certificate and CA
+  certificate.
+- Modify OVNKubernetes operator configuration object in order to allow
+  enablement of IPsec for the cluster.
+- Certificate rotation will be handled in the same as it is with the current
+  OVN TLS implementation.
+
+### Risks and Mitigations
+
+#### Upstreaming
+
+A number of changes will need to be made to upstream communities including the
+Linux kernel. As these may need time to execute, it is imperative that we
+identify them early and start the process.
```

Comment:
I did not see this coming as I read the rest of the enhancement.  What kernel change(s) are expected?

I thought ipsec support was already in OVN, so it was more just getting it enabled in ovn-kube?  https://docs.ovn.org/en/latest/tutorials/ovn-ipsec.html

If the work below ovn-kube is insufficient, can you highlight the gaps?

Reply:
There's at least one kernel patch that got applied to the ovs downstream datapath that is not in the upstream kernel. Mark is working on getting that upstream.

Reply:
https://patchwork.ozlabs.org/project/openvswitch/patch/20180919211558.13937-2-qiuyu.xiao.qyx@gmail.com/

I am unable to get OVS IPsec with Geneve or OVN IPsec working without it. I have patched the in-tree kernel module and it then works.